# Исследование надёжности заёмщиков.

<h1>Оглавление:<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Описание-проекта:" data-toc-modified-id="Описание-проекта:-0.1"><span class="toc-item-num">0.1&nbsp;&nbsp;</span>Описание проекта:</a></span></li><li><span><a href="#Описание-данных:" data-toc-modified-id="Описание-данных:-0.2"><span class="toc-item-num">0.2&nbsp;&nbsp;</span>Описание данных:</a></span></li></ul></li><li><span><a href="#Шаг-1.-Импортирую-необходимые-библиотеки,-открываю-файл-с-данными-и-изучаю-общую-информацию." data-toc-modified-id="Шаг-1.-Импортирую-необходимые-библиотеки,-открываю-файл-с-данными-и-изучаю-общую-информацию.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Шаг 1. Импортирую необходимые библиотеки, открываю файл с данными и изучаю общую информацию.</a></span><ul class="toc-item"><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Шаг-2.-Предобработка-данных." data-toc-modified-id="Шаг-2.-Предобработка-данных.-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Шаг 2. Предобработка данных.</a></span><ul class="toc-item"><li><span><a href="#Столбец-children-(количество-детей-в-семье)." data-toc-modified-id="Столбец-children-(количество-детей-в-семье).-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Столбец <code>children</code> (количество детей в семье).</a></span></li><li><span><a href="#Столбец-days_employed-(общий-трудовой-стаж-в-днях)." data-toc-modified-id="Столбец-days_employed-(общий-трудовой-стаж-в-днях).-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Столбец <code>days_employed</code> (общий трудовой стаж в днях).</a></span></li><li><span><a href="#Столбец-dob_years-(возраст-клиента-в-годах)." data-toc-modified-id="Столбец-dob_years-(возраст-клиента-в-годах).-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Столбец <code>dob_years</code> (возраст клиента в годах).</a></span></li><li><span><a href="#Столбец-education-(уровень-образования-клиента)." data-toc-modified-id="Столбец-education-(уровень-образования-клиента).-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Столбец <code>education</code> (уровень образования клиента).</a></span></li><li><span><a href="#Столбец-education_id-(идентификатор-уровня-образования)." data-toc-modified-id="Столбец-education_id-(идентификатор-уровня-образования).-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Столбец <code>education_id</code> (идентификатор уровня образования).</a></span></li><li><span><a href="#Столбец-family_status-(семейное-положение)." data-toc-modified-id="Столбец-family_status-(семейное-положение).-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Столбец <code>family_status</code> (семейное положение).</a></span></li><li><span><a href="#Столбец-family_status_id-(идентификатор-семейного-положения)." data-toc-modified-id="Столбец-family_status_id-(идентификатор-семейного-положения).-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>Столбец <code>family_status_id</code> (идентификатор семейного положения).</a></span></li><li><span><a href="#Столбец-gender-(пол-клиента)." data-toc-modified-id="Столбец-gender-(пол-клиента).-2.8"><span class="toc-item-num">2.8&nbsp;&nbsp;</span>Столбец <code>gender</code> (пол клиента).</a></span></li><li><span><a href="#Столбец-income_type-(тип-занятости)." data-toc-modified-id="Столбец-income_type-(тип-занятости).-2.9"><span class="toc-item-num">2.9&nbsp;&nbsp;</span>Столбец <code>income_type</code> (тип занятости).</a></span></li><li><span><a href="#Столбец-debt-(имел-ли-задолженность-по-возврату-кредитов)." data-toc-modified-id="Столбец-debt-(имел-ли-задолженность-по-возврату-кредитов).-2.10"><span class="toc-item-num">2.10&nbsp;&nbsp;</span>Столбец <code>debt</code> (имел ли задолженность по возврату кредитов).</a></span></li><li><span><a href="#Столбец-total_income-(ежемесячный-доход)." data-toc-modified-id="Столбец-total_income-(ежемесячный-доход).-2.11"><span class="toc-item-num">2.11&nbsp;&nbsp;</span>Столбец <code>total_income</code> (ежемесячный доход).</a></span></li><li><span><a href="#Столбец-purpose-(цель-получения-кредита)." data-toc-modified-id="Столбец-purpose-(цель-получения-кредита).-2.12"><span class="toc-item-num">2.12&nbsp;&nbsp;</span>Столбец <code>purpose</code> (цель получения кредита).</a></span></li><li><span><a href="#Удаляю-явные-дубликаты." data-toc-modified-id="Удаляю-явные-дубликаты.-2.13"><span class="toc-item-num">2.13&nbsp;&nbsp;</span>Удаляю явные дубликаты.</a></span></li><li><span><a href="#Смотрю-итоговый-результат." data-toc-modified-id="Смотрю-итоговый-результат.-2.14"><span class="toc-item-num">2.14&nbsp;&nbsp;</span>Смотрю итоговый результат.</a></span></li><li><span><a href="#Вывод:" data-toc-modified-id="Вывод:-2.15"><span class="toc-item-num">2.15&nbsp;&nbsp;</span>Вывод:</a></span></li></ul></li><li><span><a href="#Шаг-3.-Отвечаю-на-вопросы." data-toc-modified-id="Шаг-3.-Отвечаю-на-вопросы.-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Шаг 3. Отвечаю на вопросы.</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-наличием-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между наличием детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-целями-кредита-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-целями-кредита-и-возвратом-кредита-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Есть ли зависимость между целями кредита и возвратом кредита в срок?</a></span></li></ul></li><li><span><a href="#Общий-вывод:" data-toc-modified-id="Общий-вывод:-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод:</a></span></li></ul></div>

### Описание проекта:
Заказчик - кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка - статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга - специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Описание данных:
Датасет `data` содержит:
- **`children`** - количество детей в семье;
- **`days_employed`** - общий трудовой стаж в днях;
- **`dob_years`** - возраст клиента в годах;
- **`education`** - уровень образования клиента;
- **`education_id`** - идентификатор уровня образования;
- **`family_status`** - семейное положение;
- **`family_status_id`** - идентификатор семейного положения;
- **`gender`** - пол клиента;
- **`income_type`** - тип занятости;
- **`debt`** - имел ли задолженность по возврату кредитов;
- **`total_income`** - ежемесячный доход;
- **`purpose`** - цель получения кредита.

## Шаг 1. Импортирую необходимые библиотеки, открываю файл с данными и изучаю общую информацию.

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('/datasets/data.csv')

In [3]:
def df_info(df):
    display(df.head(3))
    print('-' * 50)
    df.info()
    print('-' * 50)
    display(df.describe(include = 'all'))
    print('-' * 50)
    print('Количество явных дубликатов:', df.duplicated().sum())

In [4]:
df_info(df)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   

     family_status  family_status_id gender income_type  debt   total_income  \
0  женат / замужем                 0      F   сотрудник     0  253875.639453   
1  женат / замужем                 0      F   сотрудник     0  112080.014102   
2  женат / замужем                 0      M   сотрудник     0  145885.952297   

                   purpose  
0            покупка жилья  
1  приобретение автомобиля  
2            покупка жилья

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB
--------------------------------------------------


children  days_employed     dob_years education  education_id  \
count   21525.000000   19351.000000  21525.000000     21525  21525.000000   
unique           NaN            NaN           NaN        15           NaN   
top              NaN            NaN           NaN   среднее           NaN   
freq             NaN            NaN           NaN     13750           NaN   
mean        0.538908   63046.497661     43.293380       NaN      0.817236   
std         1.381587  140827.311974     12.574584       NaN      0.548138   
min        -1.000000  -18388.949901      0.000000       NaN      0.000000   
25%         0.000000   -2747.423625     33.000000       NaN      1.000000   
50%         0.000000   -1203.369529     42.000000       NaN      1.000000   
75%         1.000000    -291.095954     53.000000       NaN      1.000000   
max        20.000000  401755.400475     75.000000       NaN      4.000000   

          family_status  family_status_id gender income_type          debt  \
count             21525      21525.000000  21525       21525  21525.000000   
unique                5               NaN      3           8           NaN   
top     женат / замужем               NaN      F   сотрудник           NaN   
freq              12380               NaN  14236       11119           NaN   
mean                NaN          0.972544    NaN         NaN      0.080883   
std                 NaN          1.420324    NaN         NaN      0.272661   
min                 NaN          0.000000    NaN         NaN      0.000000   
25%                 NaN          0.000000    NaN         NaN      0.000000   
50%                 NaN          0.000000    NaN         NaN      0.000000   
75%                 NaN          1.000000    NaN         NaN      0.000000   
max                 NaN          4.000000    NaN         NaN      1.000000   

        total_income  purpose  
count   1.935100e+04    21525  
unique           NaN       38  
top              NaN  свадьба  
freq             NaN      797  
mean    1.674223e+05      NaN  
std     1.029716e+05      NaN  
min     2.066726e+04      NaN  
25%     1.030532e+05      NaN  
50%     1.450179e+05      NaN  
75%     2.034351e+05      NaN  
max     2.265604e+06      NaN

--------------------------------------------------
Количество явных дубликатов: 54


### Вывод:
Датасет содержит **21525** записей по **12** столбцам. Типы данных на первый взгляд соответствуют содержимому столбцов, разве что столбцы **`days_employed`** и **`total_income`** имеют вещественный тип данных, а не целочисленный. Названия столбцов внесены корректно, без стилистических нарушений. В следующих столбцах содержатся пропущенные значения:
- **`days_employed`** - **2174** пропуска;
- **`total_income`** - **2174** пропуска.

В столбцах **`children`** и **`days_employed`** есть отрицательные и некорректные значения, в столбце **`dob_years`** есть некорректные значения, в столбце **`total_income`** есть слишком большие значения. В столбцах **`education`** и **`purpose`** встречаются неявные дубликаты. Также датасет содержит **54** явных дубликата. Данные изначально плохого качества и требуют предобработки.

## Шаг 2. Предобработка данных. 

Начинаю по порядку исправлять это безобразие.

### Столбец `children` (количество детей в семье).

Смотрю уникальные значения данного столбца и их колличество.

In [5]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

Несколько выбиваются значения **-1** и **20**. По поводу значения **-1**, либо анкету в банке заполняли циничные родители потерявшие ребенка, и данные так и внесли в систему, либо, что более вероятно, это техническая ошибка и подразумевалось значение **1**. А по поводу значения **20** сомнений больше, вполне возможно что кто-то имеет **20** детей, но это скорее единичный случай, а в нашем датасете **76** таких случаев. Часть детей конечно могут быть приемными, но приемных детей дают только тем семьям, которые могут их обеспечить, такие семьи как правило не берут кредиты. Сортирую датасет по заемщикам с **20** детьми, чтобы оценить показатели других столбцов и решить наверняка, что делать.

In [6]:
df.query('children == 20').head(10)

children  days_employed  dob_years education  education_id  \
606         20    -880.221113         21   среднее             1   
720         20    -855.595512         44   среднее             1   
1074        20   -3310.411598         56   среднее             1   
2510        20   -2714.161249         59    высшее             0   
2941        20   -2161.591519          0   среднее             1   
3302        20            NaN         35   среднее             1   
3396        20            NaN         56    высшее             0   
3671        20    -913.161503         23   среднее             1   
3697        20   -2907.910616         40   среднее             1   
3735        20    -805.044438         26    высшее             0   

              family_status  family_status_id gender  income_type  debt  \
606         женат / замужем                 0      M    компаньон     0   
720         женат / замужем                 0      F    компаньон     0   
1074        женат / замужем                 0      F    сотрудник     1   
2510         вдовец / вдова                 2      F    сотрудник     0   
2941        женат / замужем                 0      F    сотрудник     0   
3302  Не женат / не замужем                 4      F  госслужащий     0   
3396        женат / замужем                 0      F    компаньон     0   
3671  Не женат / не замужем                 4      F    сотрудник     0   
3697       гражданский брак                 1      M    сотрудник     0   
3735  Не женат / не замужем                 4      M    сотрудник     0   

       total_income                                purpose  
606   145334.865002                          покупка жилья  
720   112998.738649                   покупка недвижимости  
1074  229518.537004                  получение образования  
2510  264474.835577  операции с коммерческой недвижимостью  
2941  199739.941398                  на покупку автомобиля  
3302            NaN                 профильное образование  
3396            NaN                     высшее образование  
3671  101255.492076     на покупку подержанного автомобиля  
3697  115380.694664     на покупку подержанного автомобиля  
3735  137200.646181                           ремонт жилью

Мда, 23 летяя девушка с 20 детьми, вот так жизнь у человека сложилась... Это очевидно также техническая ошибка. Возможно при первом обращении в банк данные граждане имели **0** детей, а при повторном обращении, через несколько лет уже **2** детей. И замена значений произошла некорректно. Исправляю данное недоразумение.

In [7]:
df['children'] = df['children'].replace(-1, 1)
df['children'] = df['children'].replace(20, 2)
df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

### Столбец `days_employed` (общий трудовой стаж в днях).

In [8]:
df['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

In [9]:
df['days_employed'].isna().sum()

2174

Я имею, выборку от имеющих отрицательный трудовой стаж, до тех, кто работал более **1100** лет. А также **2174** пропущенных значения. Крайне занятно. В столбце **`total_income`** имеется такое же количество пропусков. Проверяю, в одних ли они строках.

In [10]:
df[(df['days_employed'].isna()) & (df['total_income'].isna())].shape[0]

2174

Так и есть. Смотрю долю пропущенных значений для обоих столбцов.

In [11]:
df['days_employed'].isna().mean()

0.10099883855981417

**10.1%** - это довольно много, в данном случае, удаление строк с пропущенными значениями может сильно повлиять на финальный результат исследования. Отталкиваясь от того, что запросить корректные и недостающие данные не представляется возможным. А так как цель исследования найти корреляцию между семейным положением и количеством детей клиента с фактом погашения кредита в срок, и напрямую не связана с трудовым стажем и ежемесячным доходом, лучшим решением будет заполнить пропуски в столбцах **`days_employed`** и **`total_income`** медианным значением. 
Приступаю к обработке столбца **`days_employed`**. Сначала, с помощью метода `abs()`, избавляюсь от отрицательных значений, а после, опираясь на тип занятости, заполняю недостающие значения, и в конце, с помощью метода `astype()`, меняю тип данных на целочисленный.

In [12]:
df['days_employed'] = abs(df['days_employed'])
df['days_employed'] = df['days_employed'].fillna(df.groupby('income_type')['days_employed'].transform('median'))
df['days_employed'] = df['days_employed'].astype('int64')
df['days_employed'].isna().sum()

0

Вуаля, бесполезные данные восстановлены. Судя по одинаковому количеству пропущенных значений, вероятно, произошел технический сбой при переносе данных из одного датасета в другой. Возможно из-за несовпадения типов данных этих столбцов.

### Столбец `dob_years` (возраст клиента в годах).

Смотрю уникальные значения данного столбца.

In [13]:
df['dob_years'].sort_values().unique()

array([ 0, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51,
       52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68,
       69, 70, 71, 72, 73, 74, 75], dtype=int64)

Возраст клиента **0** лет? Серьезно? Кредит на новую люльку? Эти банки настолько понизили возрастной порог?
Сортирую датасет по младенцам, чтобы оценить значения других столбцов.

In [14]:
df.query('dob_years == 0').head(10)

children  days_employed  dob_years education  education_id  \
99           0         346541          0   Среднее             1   
149          0           2664          0   среднее             1   
270          3           1872          0   среднее             1   
578          0         397856          0   среднее             1   
1040         0           1158          0    высшее             0   
1149         0            934          0   среднее             1   
1175         0         370879          0   среднее             1   
1386         0           5043          0    высшее             0   
1890         0           1574          0    высшее             0   
1898         0         370144          0   среднее             1   

              family_status  family_status_id gender  income_type  debt  \
99          женат / замужем                 0      F    пенсионер     0   
149               в разводе                 3      F    сотрудник     0   
270         женат / замужем                 0      F    сотрудник     0   
578         женат / замужем                 0      F    пенсионер     0   
1040              в разводе                 3      F    компаньон     0   
1149        женат / замужем                 0      F    компаньон     0   
1175        женат / замужем                 0      F    пенсионер     0   
1386        женат / замужем                 0      M  госслужащий     0   
1890  Не женат / не замужем                 4      F    сотрудник     0   
1898         вдовец / вдова                 2      F    пенсионер     0   

       total_income                                 purpose  
99     71291.522491                              автомобиль  
149    70176.435951                       операции с жильем  
270   102166.458894                            ремонт жилью  
578    97620.687042  строительство собственной недвижимости  
1040  303994.134987                         свой автомобиль  
1149  201852.430096                    покупка недвижимости  
1175  313949.845188   получение дополнительного образования  
1386  240523.618071                    сделка с автомобилем  
1890            NaN                                   жилье  
1898  127400.268338                   на покупку автомобиля

Очевидно опять техническая ошибка, БД данного банка оставляет желать лучшего. Так как возраст клиентов напрямую не влияет на цель исследования, заменяю **0** на медианное значение.

In [15]:
df['dob_years'] = df['dob_years'].replace(0, df.query('dob_years != 0')['dob_years'].median())
df['dob_years'].sort_values().unique()

array([19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35,
       36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52,
       53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69,
       70, 71, 72, 73, 74, 75], dtype=int64)

Готово. Теперь кредиты строго с 18 лет!

### Столбец `education` (уровень образования клиента).

Смотрю уникальные значения данного столбца.

In [16]:
df['education'].sort_values().unique()

array(['ВЫСШЕЕ', 'Высшее', 'НАЧАЛЬНОЕ', 'НЕОКОНЧЕННОЕ ВЫСШЕЕ',
       'Начальное', 'Неоконченное высшее', 'СРЕДНЕЕ', 'Среднее',
       'УЧЕНАЯ СТЕПЕНЬ', 'Ученая степень', 'высшее', 'начальное',
       'неоконченное высшее', 'среднее', 'ученая степень'], dtype=object)

Здесь налицо неявные дубликаты, видимо значения вносились разными работниками банка, не имеющими общей стилистики. Привожу все значения столбцов со строками к нижнему регистру.

In [17]:
columns = ['education', 'family_status', 'gender', 'income_type', 'purpose']
for column in columns:
    df[column] = df[column].str.lower()
    
df['education'].sort_values().unique()

array(['высшее', 'начальное', 'неоконченное высшее', 'среднее',
       'ученая степень'], dtype=object)

### Столбец `education_id` (идентификатор уровня образования).

Смотрю уникальные значения данного столбца.

In [18]:
df['education_id'].sort_values().unique()

array([0, 1, 2, 3, 4], dtype=int64)

Ого, неужели хоть один столбец в порядке? 5 уровней образования клиента и 5 идентификаторов уровня образования.
Выясняю, какой идентификатор, какому уровню соответствует, для этого создаю новый датасет `education`.

In [19]:
education = df[['education', 'education_id']].drop_duplicates().reset_index(drop=True)
education

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

Неплохо, никаких аномалий. Теперь столбец **`education`** не нужен, и его можно смело удалить. А значения столбца **`education_id`** всегда можно получить по идентификатору в датасете `education`.

In [20]:
df.drop('education', axis=1, inplace=True)

### Столбец `family_status` (семейное положение).

Смотрю уникальные значения данного столбца.

In [21]:
df['family_status'].sort_values().unique()

array(['в разводе', 'вдовец / вдова', 'гражданский брак',
       'женат / замужем', 'не женат / не замужем'], dtype=object)

Неужели еще один столбец, где ничего не нужно редактировать?!

### Столбец `family_status_id` (идентификатор семейного положения).

Смотрю уникальные значения данного столбца.

In [22]:
df['family_status_id'].sort_values().unique()

array([0, 1, 2, 3, 4], dtype=int64)

Здесь тоже все хорошо. Как и в столбце **`education_id`**, определяю, какой идентификатор, чему соответствует. Создаю датасет `family_status` с уникальными значениями столбцов **`family_status`** и **`family_status_id`**.

In [23]:
family_status = df[['family_status', 'family_status_id']].drop_duplicates().reset_index(drop=True)
family_status

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  не женат / не замужем                 4

Это не может не радовать, вот бы все столбцы были в таком порядке. Теперь столбец **`family_status`** также не нужен, и его можно удалить.

In [24]:
df.drop('family_status', axis=1, inplace=True)

### Столбец `gender` (пол клиента).

Смотрю уникальные значения данного столбца и их колличество.

In [25]:
df['gender'].value_counts()

f      14236
m       7288
xna        1
Name: gender, dtype: int64

In [26]:
df.query('gender == "xna"')

children  days_employed  dob_years  education_id  family_status_id  \
10701         0           2358         24             2                 1   

      gender income_type  debt   total_income               purpose  
10701    xna   компаньон     0  203905.157261  покупка недвижимости

Насколько мне известно, официально, в России нет трансгендеров, а лица сменившие пол в любом случае значаться как **f** и **m**. Данную строку можно удалить и она не повлияет на исследование, но по этой же причине её можно и оставить. Выбираю второе, да будет **xna**!

### Столбец `income_type` (тип занятости).

Смотрю уникальные значения данного столбца.

In [27]:
df['income_type'].sort_values().unique()

array(['безработный', 'в декрете', 'госслужащий', 'компаньон',
       'пенсионер', 'предприниматель', 'сотрудник', 'студент'],
      dtype=object)

Никаких дублей. Лайк.

### Столбец `debt` (имел ли задолженность по возврату кредитов).

Смотрю уникальные значения данного столбца и их колличество.

In [28]:
df['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

Обошлось без аномалий. Обозначений данных я не нашел, но исходя из количественных и логических размышлений, смею предположить что:   
- **`0`** - не имел задолженностей; 
- **`1`** - имел.

### Столбец `total_income` (ежемесячный доход).

Как я решил, когда обрабатывал столбец **`days_employed`**, так как цель исследования напрямую не связана с ежемесячным доходом, лучшим решением будет заполнить пропуски медианным значением. А также меняю тип данных на целочисленный. Благо, в этот раз обошлось без отрицательных значений.

In [29]:
df['total_income'] = df['total_income'].fillna(df.groupby('income_type')['total_income'].transform('median'))
df['total_income'] = df['total_income'].astype('int64')
df['total_income'].isna().sum()

0

На основании диапазонов, указанных ниже, создаю столбец **`total_income_category`** с категориями дохода:
- **до 30000** - **`'E'`**;
- **30001 - 50000** - **`'D'`**;
- **50001 - 200000** - **`'C'`**;
- **200001 - 1000000** - **`'B'`**;
- **1000001 и выше** - **`'A'`**.

In [30]:
def total_income_values(total_income):
    try:
        if total_income <= 30000:
            return 'e'

        elif 30001 <= total_income <= 50000:
            return 'd'

        elif 50001 <= total_income <= 200000:
            return 'c'

        elif 200001 <= total_income <= 1000000:
            return 'b'
        
        elif 1000001 <= total_income:
            return 'a'
        
        else:
            return 'unknown'
    except:
        print('Что-то пошло не так, перепроверь код.')

df['total_income_category'] = df['total_income'].apply(total_income_values)
df['total_income_category'].value_counts()

c    16086
b     5042
d      350
a       25
e       22
Name: total_income_category, dtype: int64

Доходы конечно не среднероссийские. Но мне даже хочется верить, что это так, а не очередная ошибка в данных.

### Столбец `purpose` (цель получения кредита).

Смотрю уникальные значения данного столбца.

In [31]:
df['purpose'].sort_values().unique()

array(['автомобили', 'автомобиль', 'высшее образование',
       'дополнительное образование', 'жилье',
       'заняться высшим образованием', 'заняться образованием',
       'на покупку автомобиля', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля', 'на проведение свадьбы',
       'недвижимость', 'образование', 'операции с жильем',
       'операции с коммерческой недвижимостью',
       'операции с недвижимостью', 'операции со своей недвижимостью',
       'покупка жилой недвижимости', 'покупка жилья',
       'покупка жилья для сдачи', 'покупка жилья для семьи',
       'покупка коммерческой недвижимости', 'покупка недвижимости',
       'покупка своего жилья', 'получение высшего образования',
       'получение дополнительного образования', 'получение образования',
       'приобретение автомобиля', 'профильное образование',
       'ремонт жилью', 'свадьба', 'свой автомобиль',
       'сделка с автомобилем', 'сделка с подержанным автомобилем',
       'строительство 

Вновь море неявных дубликатов. Избавляюсь от них функцией `purpose_values`. Создаю столбец **`purpose_category`**, куда записываю отсортированные значения. А столбец **`purpose`** за ненадобностью удаляю.

In [32]:
def purpose_values(purpose):
    try:
        if 'свадьб' in purpose:
            return 'проведение свадьбы'

        elif 'образован' in purpose:
            return 'получение образования'

        elif 'автомобил' in purpose:
            return 'операции с автомобилем'
        
        elif 'жиль' in purpose or 'недвижимост' in purpose:
            return 'операции с недвижимостью'

        else:
            return 'unknown'
    except:
        print('Что-то пошло не так, переповерь код.')

df['purpose_category'] = df['purpose'].apply(purpose_values)
df.drop('purpose', axis=1, inplace=True)
df['purpose_category'].value_counts()

операции с недвижимостью    10840
операции с автомобилем       4315
получение образования        4022
проведение свадьбы           2348
Name: purpose_category, dtype: int64

Отлично, с этим уже можно работать.

### Удаляю явные дубликаты.

Проверяю не добавилось ли явных дубликатов в датасете.

In [33]:
print(df.duplicated().sum())
print(f'Доля явных дубликатов составляет {df.duplicated().sum() / df.shape[0]:.1%}')

408
Доля явных дубликатов составляет 1.9%


До предобработки датасет содержал **54** явных дубликата, теперь их **408**.

In [34]:
df = df.drop_duplicates().reset_index(drop=True)
df.duplicated().sum()

0

### Смотрю итоговый результат.

In [35]:
df_info(df)

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1           8437         42             0                 0      f   
1         1           4024         36             1                 0      f   
2         0           5623         33             1                 0      m   

  income_type  debt  total_income total_income_category  \
0   сотрудник     0        253875                     b   
1   сотрудник     0        112080                     c   
2   сотрудник     0        145885                     c   

           purpose_category  
0  операции с недвижимостью  
1    операции с автомобилем  
2  операции с недвижимостью

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21117 entries, 0 to 21116
Data columns (total 11 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   children               21117 non-null  int64 
 1   days_employed          21117 non-null  int64 
 2   dob_years              21117 non-null  int64 
 3   education_id           21117 non-null  int64 
 4   family_status_id       21117 non-null  int64 
 5   gender                 21117 non-null  object
 6   income_type            21117 non-null  object
 7   debt                   21117 non-null  int64 
 8   total_income           21117 non-null  int64 
 9   total_income_category  21117 non-null  object
 10  purpose_category       21117 non-null  object
dtypes: int64(7), object(4)
memory usage: 1.8+ MB
--------------------------------------------------


children  days_employed     dob_years  education_id  \
count   21117.000000   21117.000000  21117.000000  21117.000000   
unique           NaN            NaN           NaN           NaN   
top              NaN            NaN           NaN           NaN   
freq             NaN            NaN           NaN           NaN   
mean        0.483923   66433.166169     43.412038      0.815978   
std         0.758165  138666.364793     12.213994      0.551518   
min         0.000000      24.000000     19.000000      0.000000   
25%         0.000000    1010.000000     33.000000      1.000000   
50%         0.000000    2018.000000     43.000000      1.000000   
75%         1.000000    5302.000000     53.000000      1.000000   
max         5.000000  401755.000000     75.000000      4.000000   

        family_status_id gender income_type          debt  total_income  \
count       21117.000000  21117       21117  21117.000000  2.111700e+04   
unique               NaN      3           8           NaN           NaN   
top                  NaN      f   сотрудник           NaN           NaN   
freq                 NaN  13922       10898           NaN           NaN   
mean            0.981389    NaN         NaN      0.082351  1.657240e+05   
std             1.424319    NaN         NaN      0.274905  9.889210e+04   
min             0.000000    NaN         NaN      0.000000  2.066700e+04   
25%             0.000000    NaN         NaN      0.000000  1.068060e+05   
50%             0.000000    NaN         NaN      0.000000  1.425940e+05   
75%             1.000000    NaN         NaN      0.000000  1.969830e+05   
max             4.000000    NaN         NaN      1.000000  2.265604e+06   

       total_income_category          purpose_category  
count                  21117                     21117  
unique                     5                         4  
top                        c  операции с недвижимостью  
freq                   15678                     10575  
mean                     NaN                       NaN  
std                      NaN                       NaN  
min                      NaN                       NaN  
25%                      NaN                       NaN  
50%                      NaN                       NaN  
75%                      NaN                       NaN  
max                      NaN                       NaN

--------------------------------------------------
Количество явных дубликатов: 0


### Вывод:
Я заполнил пропуски, проверил данные на аномалии и исправил их, измененил некорректные типы данных, сформировал дополнительные датасеты-словари, произвел декомпозицию исходного датасета, категоризировал доход и цели кредита, а также удалил дубликаты.

## Шаг 3. Отвечаю на вопросы. 

### Есть ли зависимость между наличием детей и возвратом кредита в срок?
Для этого собираю данные из столбцов **`children`** и **`debt`** в датасете `children_credit` и считаю количество  должников (`sum_debt`), общее количество клиентов (`count_debt`) и долю должников от общего количества (`mean_debt`).

In [36]:
children_credit = df.pivot_table(index = 'children', values = 'debt', aggfunc = ['sum', 'count', 'mean'])
children_credit.columns = ['sum_debt', 'count_debt', 'mean_debt']
children_credit

sum_debt  count_debt  mean_debt
children                                 
0             1061       13827   0.076734
1              445        4797   0.092766
2              202        2115   0.095508
3               27         329   0.082067
4                4          40   0.100000
5                0           9   0.000000

Можно сделать вывод, что небольшая зависимость между наличием детей и возвратом кредита в срок есть, а вот количество детей особо не играет роли. Группы с **1**, **2** и **3** детьми имеют большую вероятность стать должниками, чем группа без детей. Группы с **4** и **5** детьми, за малым количеством мы в рассчет не берем.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?
Для этого собираю данные из столбцов **`family_status_id`** и **`debt`** в датасете `family_credit` и считаю количество  должников (`sum_debt`), общее количество клиентов (`count_debt`) и долю должников от общего количества (`mean_debt`).

In [37]:
family_credit = df.pivot_table(index = 'family_status_id', values = 'debt', aggfunc = ['sum', 'count', 'mean'])
family_credit.columns = ['sum_debt', 'count_debt', 'mean_debt']
family_credit

sum_debt  count_debt  mean_debt
family_status_id                                 
0                      929       12074   0.076942
1                      388        4123   0.094106
2                       63         943   0.066808
3                       85        1193   0.071249
4                      274        2784   0.098420

Можно сделать вывод, что небольшая зависимость между семейным положением и возвратом кредита в срок есть. Заемщики, не состоящие в браке (**"гражданский брак"**, **"не женат / не замужем"**) наиболее склонны к задержке выплат, в то время как **"вдовцы / водвы"** меньше всех склонны к задолженностям по кредиту.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
Для этого собираю данные из столбцов **`total_income_category`** и **`debt`** в датасете `income_credit` и считаю количество  должников (`sum_debt`), общее количество клиентов (`count_debt`) и долю должников от общего количества (`mean_debt`).

In [38]:
income_credit = df.pivot_table(index = 'total_income_category', values = 'debt', aggfunc = ['sum', 'count', 'mean'])
income_credit.columns = ['sum_debt', 'count_debt', 'mean_debt']
income_credit

sum_debt  count_debt  mean_debt
total_income_category                                 
a                             2          25   0.080000
b                           356        5042   0.070607
c                          1358       15678   0.086618
d                            21         350   0.060000
e                             2          22   0.090909

Количество заемщиков категорий **A** и **C**, на мой взгляд, недостаточно, для анализа. К тому же, в столбце с доходом изначально были сомнительно высокие показатели, не удивительно, что результат анализа тоже странный. Будь это основной целью исследования, следовало бы заменить все завышенные значения на медианные, опираясь на такие показатели как пол, возраст и занимаемая должность. Но имеем, что имеем.

Можно сделать вывод, что небольшая зависимость между уровнем дохода и возвратом кредита в срок есть. Наибольшая вероятность стать должником у заемщиков, из категории **E** (с заработком **до 30000**). А наименьшая у заемщиков из категории **D** (с заработком **от 30001 до 50000**).

### Есть ли зависимость между целями кредита и возвратом кредита в срок?
Для этого собираю данные из столбцов **`purpose_category`** и **`debt`** в датасете `purpose_credit` и считаю количество  должников (`sum_debt`), общее количество клиентов (`count_debt`) и долю должников от общего количества (`mean_debt`).

In [39]:
purpose_credit = df.pivot_table(index = 'purpose_category', values = 'debt', aggfunc = ['sum', 'count', 'mean'])
purpose_credit.columns = ['sum_debt', 'count_debt', 'mean_debt']
purpose_credit

sum_debt  count_debt  mean_debt
purpose_category                                         
операции с автомобилем         402        4272   0.094101
операции с недвижимостью       781       10575   0.073853
получение образования          370        3964   0.093340
проведение свадьбы             186        2306   0.080659

 Можно сделать вывод, что небольшая зависимость между целями кредита и возвратом кредита в срок есть. Наибольшая вероятность стать должником у заемщиков, получающих кредит на **"операции с автомобилем"**.

## Общий вывод:

Полученные от банка данные имели большое количество пропусков и аномалий, предположительно результат технических ошибок. Используя логику и функции я заполнил пропуски, и исправили аномалии данных, а также измененил некорректные типы данных. Сформировал дополнительные датасеты-словари, произвел декомпозицию исходного датасета, категоризировал доход и цели кредита а также удалил дубликаты. Данные стали пригодными для анализа, чем я и занялся. 
***
Я собрал требуемые для анализа данные в сводные таблици и в результате, основываясь на полученных данных можно сделать вывод, что семейное положение и наличие детей в семье влияют на факт погашения кредита. Вероятность заемщика принадлежащего к группе **"без детей"** стать должником ниже, чем у заемщиков из групп **"с детьми"**, вероятность заемщика из группы **"не женат / не замужем"** выше, чем у заемщиков остальных групп. А заодно я выяснил что наибольшая вероятность стать должником есть у заемщиков, с заработком **до 30000**. А наименьшая у заемщиков с заработком **от 30001 до 50000**. И, также, наибольшая вероятность стать должником у заемщиков, получающих кредит на **"операции с автомобилем"**. 